OpenRA.Game/Server/Server.cs

```diff
@@ -559,24 +559,32 @@ public void DropClient(Connection toDrop)
 				// Send disconnected order, even if still in the lobby
 				DispatchOrdersToClients(toDrop, 0, new ServerOrder("Disconnected", "").Serialize());
 
-				LobbyInfo.Clients.RemoveAll(c => c.Index == toDrop.PlayerIndex);
-				LobbyInfo.ClientPings.RemoveAll(p => p.Index == toDrop.PlayerIndex);
-
-				// Client was the server admin
-				// TODO: Reassign admin for game in progress via an order
-				if (Dedicated && dropClient.IsAdmin && State == ServerState.WaitingPlayers)
+				// Client was the server admin, but the game isn't over yet.
+				if (Dedicated && dropClient.IsAdmin)
 				{
-					// Remove any bots controlled by the admin
-					LobbyInfo.Clients.RemoveAll(c => c.Bot != null && c.BotControllerClientIndex == toDrop.PlayerIndex);
-
-					var nextAdmin = LobbyInfo.Clients.Where(c1 => c1.Bot == null)
-						.MinByOrDefault(c => c.Index);
+					// Pick the human player who joined right after the previous admin.
+					var nextAdmin = LobbyInfo.Clients.Where(c => c.Bot == null && c != dropClient).MinByOrDefault(c => c.Index);
 
-					if (nextAdmin != null)
+					// Still in the lobby so handle things directly.
+					if (State == ServerState.WaitingPlayers)
 					{
-						nextAdmin.IsAdmin = true;
-						SendMessage("{0} is now the admin.".F(nextAdmin.Name));
+						// Remove the previous admin.
+						LobbyInfo.Clients.RemoveAll(c => c.Index == toDrop.PlayerIndex);
+
+						// Stop pinging when a player drops from the server.
+						LobbyInfo.ClientPings.RemoveAll(p => p.Index == toDrop.PlayerIndex);
+
+						// Remove any bots controlled by the previous admin.
+						LobbyInfo.Clients.RemoveAll(c => c.Bot != null && c.BotControllerClientIndex == toDrop.PlayerIndex);
+
+						if (nextAdmin != null)
+						{
+							nextAdmin.IsAdmin = true;
+							SendMessage("{0} is now the admin.".F(nextAdmin.Name));
+						}
 					}
+					else if (nextAdmin != null)
+						DispatchOrdersToClients(toDrop, 0, Order.NewAdmin(nextAdmin.Index).Serialize());
```

Comment:
Could you clarify something for me? Is the `orderManager.LobbyInfo.Clients.RemoveAll(c => c.Bot != null && c.BotControllerClientIndex == client.Index);` line that runs if no admin can be found when resolving the order important?

If so, should we not send the order (with some dummy index) anyway to cause it to run?
If not, why do we need to run it if the server found an admin and sent the order, but the client fails to resolve the index?

Just trying to straighten out the logic in my head here.

Reply:
I believe that the only case where `nextAdmin` can be null is when there are no players left in the server.  In that case it doesn't matter what we do, the server will be shut down and recycled immediately after this.